# Operators

# Arithmetic operators

#### + -> Addition
#### - -> Subtraction
#### * -> Multiplication
#### / -> Division 
#### % -> Modulo
#### // -> Floor Division
#### ** -> Power 

In [2]:
a,b = 10,2
print(a+b)
print(a-b)
print(a*b)
print(a/b)
print(a%b)
print(a//b)
print(a**b)

12
8
20
5.0
0
5
100


In [3]:
a,b = 10.5,2
print(a+b)
print(a-b)
print(a*b)
print(a/b)
print(a%b)
print(a//b)
print(a**b)

12.5
8.5
21.0
5.25
0.5
5.0
110.25


# The difference between / and //

- / (Division) operator always performs floating-point arithmetic. it always returns the float value.
- // (Floor Division) performs both floating point and integral arithmetic. If arguments are of type int, then the result type is int. If one argument is a float, then the result is a float type

## Note
- we can use +, * operators for str type also.
- If we want to use + operator for str type then compulsory both arguments should be str otherwise we will get error

In [4]:
'abc'+'10'

'abc10'

In [5]:
'abc' + 10

TypeError: can only concatenate str (not "int") to str

#### If we use the * operator for str, one argument should be int and the other argument should be str

In [6]:
'abc'*10

'abcabcabcabcabcabcabcabcabcabc'

In [7]:
'abc' * 'def'

TypeError: can't multiply sequence by non-int of type 'str'

# Relational operators
### >,>=,<,<=

In [8]:
a,b = 10,20
print("a > b is",a>b)
print("a >= b is",a>=b)
print("a < b is",a<b)
print("a <= b is",a<=b)

a > b is False
a >= b is False
a < b is True
a <= b is True


In [11]:
# We can apply relational operators for the str type also
s1,s2 = 'ashok','data'
print("a > b is",a>b)
print("a >= b is",a>=b)
print("a < b is",a<b)
print("a <= b is",a<=b)

a > b is False
a >= b is False
a < b is True
a <= b is True


In [12]:
print(True > True)
print(10>True)
print(False > True)

False
True
False


In [13]:
a = int(input('Enter 1st number:'))
b = int(input('Enter 2cd number:'))
if a >b:
    print("a is greater than b")
else:
    print("b is grater than a")

Enter 1st number: 10
Enter 2cd number: 20


b is grater than a


## Note

#### Chaining of relational operators is possible. In the chaining, all comparisons return True; then only result is True. at least one comparison returns a false result is False

In [14]:
print(10<20)
print(10<20<30)
print(10<20<30<40>50)

True
True
False


# Equality Operators 
#### ==, !=

In [16]:
print(10 == 20)
print(10 != 20)
print('ashok' == 'ashok')
print(True == False)

False
True
True
False


#### Chaining of Equality operators is possible. In the chaining, all comparisons return True; then only result is True. at least one comparison returns a false result is False.

# Logical Operators

- and -> if both arguments are True, then only result is True
- or -> if one argument is True, then the result is True
- not -> Complement

- 0 means False
- non-zero means False
- empty string is always treated as False

### x and y
- if x evalutes to False then result is x other wise y

In [19]:
20 and 10

10

In [20]:
0 and 20

0

In [22]:
'' and 20

''

## x or y
- if x evaluates to True, then result x otherwise result is y

In [23]:
print(0 or 20)

20


In [25]:
print(10 or 20)

10


In [26]:
print('' or 20)

20


## not x
- if x is evalutated to False then result is True otherwise False.

In [27]:
not 10

False

In [28]:
not ''

True

In [29]:
not 0

True

# Assignment Operators

### +=, -=,*=, /=, %=, //=, **=, &=, |=, ^=, >>=, <==

In [35]:
x = 10
x += 20
x

30

In [40]:
x = 10
x = x+20
x

30

In [38]:
x = 10
x -= 20
x

-10

In [39]:
x = 10
x = x-20
x

-10

# Tenary Operator or Conditional Operator

### syntax x = firstValue if condition else secondValue.
### If condition is True then firstValue will be considered else secondValue will be considered.

In [41]:
a = int(input('Enter 1st number: '))
b = int(input('Enter 2cd number: '))
min = a if a<b else b
min

Enter 1st number:  10
Enter 2cd number:  20


10

In [43]:
a = int(input('Enter 1st number: '))
b = int(input('Enter 2cd number: '))
c = int(input('Enter 3rd number: '))
max = a if a>b and a>c else b if b>c else c
max

Enter 1st number:  10
Enter 2cd number:  20
Enter 3rd number:  30


30

In [44]:
a = int(input('Enter 1st number: '))
b = int(input('Enter 2cd number: '))
c = int(input('Enter 3rd number: '))
min = a if a<b and a<c else b if b<c else c
min

Enter 1st number:  10
Enter 2cd number:  20
Enter 3rd number:  30


10

# Special Operators

## Identity Operators

#### 1. We can use Identity operators for address comparison.
#### 2. two types
        - is
        - is not
- r1 is r2 returns True. If both r1 and r2 are pointing to the same object.
- r1 is not r2 returns True. if both r1 and r2 are pointing to the same object.

In [47]:
a,b = 10,10
a is b

True

In [48]:
a,b = 10,20
a is b

False

In [49]:
a is not b

True

### Note: 
### What is the difference between is and ==
- is used to address comparison
- == used to content comparison

## Membership Operators

#### We can use Membership Operators to check whether the given object is present in the given collection (str, list, tuple, dict, set)
- --> in -> Returns True if the given object is present in the specified collection.
- --> not in -> Returns True if the given object is not present in the specified collection.

In [50]:
x = 'Hello learning python is very easy'
'H' in x

True

In [51]:
'i' in x

True

In [53]:
'z' in x

False

In [54]:
'z' not in x

True

In [55]:
l1 = [1,2.5,10+20j]
l2 = [3,2.5,20,True]
l1 in l2

False

In [56]:
1 in l1

True

In [57]:
'abc' in l1

False

In [58]:
'abc' not in l1

True

## Bit-wise operators 

- Decimal form(base-10)
    - The allowed digits are 0 to 9
- Binary form (base-2)
    - The allowed digits are 0 to 1

### Converting decimal to binary

7 is converting to binary

```markdown
| Base | Value | reminder |
|------|-------|--------- |
| 2    | 7     | 1        |
| 2    | 3     | 1        |
|      | 1     |          |


In [62]:
bin(7)

'0b111'

In [63]:
bin(4)

'0b100'

- 4 is converting to binary

```markdown
| Base | Value | reminder |
|------|-------|--------- |
| 2    | 4     | 0        |
| 2    | 1     | 0        |
|      | 1     |          |


### 🔢 Binary to Decimal Conversion

Convert `0b1011` to decimal:

1*(2³) + 0*(2²) + 1*(2¹) + 1*(2⁰)  
= 8 + 0 + 2 + 1  
= 11


In [80]:
int(0b1101)

13

Convert `0b1101` to decimal:

1*(2³) + 1*(2²) + 0*(2¹) + 1*(2⁰)  
= 8 + 0 + 4 + 1  
= 13

### Octal form (base-8)
    - The allowed digits are 0-7
    - It should be prefixed 0o (zero o (alpabet o))

In [81]:
x = 0o123
print(type(x))

<class 'int'>


- int to oct form

In [116]:
int(0o111)

73

- oct to int

In [113]:
oct(140)

'0o214'

## Hexa decimal form(base-16)
- It allowed digits to be 0 to 9, a - f (both lower case and upper case)
- The literal value should be prefixed with 0x
- a - 10
- b - 11
- c - 12
- d - 13
- e - 14
- f - 15

''''
0x123FACE  
= 1×16⁶ + 2×16⁵ + 3×16⁴ + 15×16³ + 10×16² + 12×16¹ + 14×16⁰  
= 16777216 + 2097152 + 196608 + 98304 + 2560 + 192 + 14  
= 190887726 '''

In [120]:
int(0x123face)
x

19135182

In [124]:
hex(12345)

'0x3039'

### Bit-wise operator

- & both bits are 1 then the result is 1
- | if at least one bit 1 then the result is 1
- ^ both bits are different then result is 1

In [126]:
bin(4)

'0b100'

In [127]:
bin(5)

'0b101'

In [128]:
5 & 4

4

In [130]:
6 & 7

6

In [129]:
print(bin(6))
print(bin(7))

0b110
0b111


In [131]:
4 | 5

5

In [133]:
7 | 8

15

In [132]:
print(bin(7))
print(bin(8))

0b111
0b1000


- for 7 we got 0b111 and for 8 we got 0b1000. for we have to take values for 7 is 111 and o is 10000. we we are applying bitwise operator we need to add zero at starting position

In [ ]:
0111
1000
====
1111

In [134]:
int(0b1111)

15

In [135]:
print(4^5)

1


In [136]:
bin(6)

'0b110'

In [137]:
bin(4)

'0b100'

In [138]:
~64

-65

# Step-by-step: `~7` (Bitwise Complement)

**Quick result:**  
`~7 = -8`  
Formula: `~n = -(n + 1)`

## Step 1 — Represent 7 in binary (8 bits)
### Decimal: 7
### Binary : 0000 0111

## Step 2 — Apply bitwise complement (~)
Flip each bit (0 → 1, 1 → 0):

- Decimal: 7
- Binary : 0000 0111
- After ~ :
- Binary : 1111 1000

`1111 1000` starts with `1` → negative number in two’s complement.

## Step 3 — Convert `1111 1000` to decimal

### Step 3.1 — Invert bits


- Binary : 1111 1000
- Invert : 0000 0111

#### Step 3.2 — Add 1
- 0000 0111
-             1
- ---------
- 0000 0110
- ---------

In [139]:
~7

-8

## 1. Left Shift (`<<`)
- The left shift operator moves all bits to the **left** by the specified number of positions.  
- Each left shift by `k` positions multiplies the number by `2^k`.
- **Formula:**
- n << k = n × (2^k)

In [141]:
5<<6

320

In [145]:
5*(2**6)

320

## 2. Right Shift (`>>`)
The right shift operator moves all bits to the **right** by the specified number of positions.  
Each right shift by `k` positions divides the number by `2^k` (integer division).

**Formula:**  
- n >> k = n // (2^k)

### Example: 20 >> 1


In [146]:
20//(2**1)

10

In [147]:
20>>1

10

# Operator Precedence
### if multiple operators are present, then which operator will be evaluated first is decided by operator precedence
-  () -> Parenthesis
-  ** -> Exponential
-  *, /, %, // -> Multiplication, Divison, Modulo, Floor Division
-  +, - -> Addition, Subtraction
-  <<, >> -> Left and right shift
-  & -> Bitwise AND
-  ^ -> Bitwise X-OR
-  | -> Bitwise or
-  >, >=, <, <=, ==, != -> Relational or Comparison
-  =, +=, -=, *= -> Assignment 
-  is, is not -> Identity
-  in, not in -> Membership
-  not -> logical not
-  and -> logical and
-  or -> logical or

In [149]:
4 ^ 5

1